src/service/resources-impl.js

```diff
@@ -407,6 +410,12 @@ export class Resources {
    * @private
    */
   buildReadyResources_() {
+    // Avoid cases where elements add more elements inside of them
+    // and cause an infinite loop of building - see #3354 for details.
+    if (this.isCurrentlyBuildingPendingResources_) {
+      return;
+    }
+    this.isCurrentlyBuildingPendingResources_ = true;
```

Comment:
The try-finally you were suggesting was for the the full for-loop right? But if the `resource.build()` step keeps failing we're never taking the resource from the `this.pendingBuildResources_` array. So next time the loop is going to run same thing going to happen, no?